doc/tutorials/platformer/step_5.md

```diff
@@ -0,0 +1,329 @@
+        ? 1
+        : 0;
+
+    return true;
+  }
+```
+
+Let's make Ember move by adding a few lines of code and creating our `update` method.  First, we
+need to define a velocity variable for Ember.  Add the following at the top of your class:
+
+```dart
+final Vector2 velocity = Vector2.zero();
+final double moveSpeed = 200;
+```
+
+This establishes a base velocity of 0 and stores `moveSpeed` so we can adjust as necessary to suit
+how the game-play should be. Next, add the `update` method with the following:
+
+
+```dart
+  @override
+  void update(double dt) {
+    velocity.x = horizontalDirection * moveSpeed;
+    position += velocity * dt;
+    super.update(dt);
+  }
+```
+
+If you run the game now, Ember moves left and right using the arrow keys or the `A` and `D` keys.
+You may have noticed that Ember doesn't look back if you are going left, to fix that, add the
+following code at the end of your `update` method:
+
+```dart
+if (horizontalDirection < 0 && scale.x > 0) {
+    flipHorizontally();
+} else if (horizontalDirection > 0 && scale.x < 0) {
+    flipHorizontally();
+}
+```
+
+Now Ember looks in the direction they are traveling.
+
+
+## Collisions
+
+It is time to get into the thick of it with collisions.  I highly suggest reading the
+[documentation](../../flame/collision_detection.md) to understand how collisions work in Flame.  The
+first thing we need to do is make the game aware that collisions are going to occur using the
+`HasCollisionDetection` mixin.  Add that to `lib/ember_quest.dart` like:
+
+```dart
+class EmberQuestGame extends FlameGame
+    with HasCollisionDetection, HasKeyboardHandlerComponents {
+```
+
+Next, add the `CollisionCallbacks` mixin to `lib/actors/ember.dart` like:
+
+```dart
+class EmberPlayer extends SpriteAnimationComponent
+    with KeyboardHandler, CollisionCallbacks, HasGameRef<EmberQuestGame> {
+```
+
+If it did not auto-import, you will need the following:
+
+```dart
+import 'package:flame/collisions.dart';
+```
+
+Now add the following `onCollision` method:
+
+```dart
+@override
+void onCollision(Set<Vector2> intersectionPoints, PositionComponent other) {
+    if (other is GroundBlock || other is PlatformBlock) {
+        if (intersectionPoints.length == 2) {
+        // Calculate the collision normal and separation distance.
+        final mid = (intersectionPoints.elementAt(0) +
+                intersectionPoints.elementAt(1)) /
+            2;
+
+        final collisionNormal = absoluteCenter - mid;
+        final separationDistance = (size.x / 2) - collisionNormal.length;
+        collisionNormal.normalize();
+
+        // If collision normal is almost upwards,
+        // ember must be on ground.
+        if (fromAbove.dot(collisionNormal) > 0.9) {
+            isOnGround = true;
+        }
+
+        // Resolve collision by moving ember along
+        // collision normal by separation distance.
+        position += collisionNormal.scaled(separationDistance);
+        }
```

Comment:
The indentation here is off. And I don't think you need to check whether the intersections points are exactly two, just do something like:
```dart
final mid = intersectionsPoints.fold(Vector2.zero(), (sum, point) => sum..add(point));
```

Reply:
I don't know, `fold()` in an entry-level tutorial? That would just confuse many of the readers.

Reply:
Indentation fixed.   The suggested line of code throws a lot of errors.  

```
Missing variable type for 'mid'.
Try adding an explicit type, or remove implicit-dynamic from your analysis options file.dart
Undefined name 'intersectionsPoints'.
Try correcting the name to one that is defined, or defining the name.dart
Avoid method calls or property accesses on a "dynamic" target.
```

I have been working it out, but still can't resolve them all.

Reply:
Ooh that's true, about the fold, didn't think about that it was a tutorial when I wrote the comment.